Further stac conversion improvements, tighten public api

jeremyh opened the pull request:
Expansion of the stac converter so it can be used from other projects.

- Use the stac `created` field instead of ODC's field name.
- Include dataset label as a 'title' property (perhaps there's a better option?)
- Separate the LS8 instruments, following the [spec example](https://github.com/radiantearth/stac-spec/blob/4988a40f6071666d345e4a1c5adc9a6338973769/item-spec/examples/landsat8-sample.json#L48-L51)
- Fall-back to WKT2 when dataset crs is not an epsg code.
- Allow custom URLs, location to be used for the output paths.

- Make all functions private except the main ones. Keep the public api small.
- Add a link to the parent stac collection in Explorer

jeremyh committed:
Tighten the public stac conversion methods, expand them

jeremyh committed:
Add WKT2 fallback for EO3 CRSes

jeremyh committed:
Only declare stac extensions that we use

sachitrajbhandari commented on eodatasets3/scripts/tostac.py:
```diff
@@ -163,74 +181,88 @@ def run(
 
 
-def dc_to_stac(
+def dataset_as_stac_item(
```
This `dc_to_stac` function is used in Landsat C3 sync script and renaming the function name and input parameters (`input_metadata_url, output_url`) will break the code.

jeremyh replied:
Oh bummer, I didn't realise it was used as a Python API already. I'll add backwards compat.

jeremyh replied:
Added shim fallback functions

sachitrajbhandari commented on eodatasets3/scripts/tostac.py:
```diff
@@ -163,74 +181,88 @@ def run(
     item_doc = dict(
         stac_version="1.0.0-beta.2",
-        stac_extensions=["eo", "projection", "view"],
+        stac_extensions=["eo", "projection"],
```
Extension `view` is used in created STAC doc after EO3 to STAC mapping. e.g.  `"eo:sun_azimuth": "view:sun_azimuth",  "eo:sun_elevation": "view:sun_elevation"`

jeremyh replied:
I'm adding it below whenever at least one property uses `view`:

https://github.com/GeoscienceAustralia/eo-datasets/pull/98/files#diff-d9f5af2b32d41023013e97e5e6b86e646f6f4cd557c3b2ba32324bcfbfa98273R262-R264

My problem is that stac items fail validation if we declare `view` but don't have the required fields, and only some datasets add them.

sachitrajbhandari reviewed (commented):
Nice refactoring with limited public functions. Does changes in `created` field and `instruments` values will required updating existing STAC doc and metadata indexed in ODC DB, any thoughts?

jeremyh commented:
> Does changes in created field and instruments values will required updating existing STAC doc and metadata indexed in ODC DB, any thoughts?

I assume this shouldn't affect ODC-indexed datasets, as the original eo3 doc is unchanged. I don't know enough about our current use of stac outputs to know if the other change matters. Will it break anything that you know? 

That seems to be the common way they're represented in the stac samples I've seen, and how the instruments field is represented on Explorer's current stac output.

sachitrajbhandari commented:
> > Does changes in created field and instruments values will required updating existing STAC doc and metadata indexed in ODC DB, any thoughts?
> 
> I assume this shouldn't affect ODC-indexed datasets, as the original eo3 doc is unchanged. I don't know enough about our current use of stac outputs to know if the other change matters. Will it break anything that you know?
> 
> That seems to be the common way they're represented in the stac samples I've seen, and how the instruments field is represented on Explorer's current stac output.

Before indexing, we do STAC to DC metadata transform. So, we will need to update transform code for changes in `created` field and `instruments` values (https://github.com/opendatacube/odc-tools/blob/develop/libs/index/odc/index/stac.py)

alexgleith commented:
I think we should consider moving this tool out of EO Datasets and into the `odc-tools` repo here: https://github.com/opendatacube/odc-tools/blob/develop/libs/index/odc/index/stac.py

Then it can sit next to the STAC to ODC conversion script linked above.

jeremyh commented:
> I think we should consider moving this tool out of EO Datasets and into the `odc-tools` repo here: https://github.com/opendatacube/odc-tools/blob/develop/libs/index/odc/index/stac.py
> 
> Then it can sit next to the STAC to ODC conversion script linked above.


We might need a wider discussion around how we're handling libraries. eo-datasets is basically all metadata handling/validation/conversion, so if this should be moved maybe all of it should? But ODC-Tools is advertising itself as experimental code and everything we have that uses eo-datasets wants extreme stability (no accidental metadata or output changes for products<sup>1</sup>). The ODC-Tools repo doesn't run its tests or have docs setup, or versioning, and isn't installable from normal pypi like eodatasets, so I'm hesitant to rely on it from other repos at the moment... (should I be?)<sup>2</sup>

This PR is exposing functions for use by Explorer's stac 1.0.0 support (and fixing the validation failures), so it's going to depend on its stability quite heavily too. 

<sup>1</sup> eo-datasets came about years ago from frustration with our metadata code, as they were diffuse and inconsistent, and slight mistakes were accidentally introduced multiple times without people noticing for months on produced products. Its tests+versioning are strict.
<sup>2</sup> I'm completely biased of course, but short-term I'd rather move stac-to-eo3 here rather than the other way around if we want them side-by-side as we have all the infrastructure set up here (and other projects already depend on eodatasets through pypi). Or else move both of them up to odc-core, but that would slow down the release cycle (?). This is the only blocker on Explorer's stac 1.0 PR  so I'm excited by any solutions that are easy :slightly_smiling_face:

jeremyh committed:
Move new stac api into a library, restore old api

The old dc_to_stac function is used externally so we need to keep
a shim of it for backwards compatibilty.

jeremyh committed:
Clearer properties, docs, for stac api

jeremyh committed:
More robust Stac schema downloads

- Use one session for all downloads
- Use a timeout
- Keep cached items much longer when they're stable Stac versions.
- Allow skipping of the cache
- Allow no-network-access usage
- Verbose option for users

jeremyh committed:
Use stac location for measurements in legacy eo3-to-stac

jeremyh commented:
Added a shim `dc_to_stac` function in the old place to maintain backwards compatibility.

I've also moved the API to `eodataset3.stac`, as the `eodatasets3.script` directory is more for CLI-frontends, not library APIs.

So it's now `eodatasets3.stac.to_stac_item()`

Added docstrings etc too, and have tried to rename the parameters to be more understandable.

jeremyh commented:
- Added a `verbose` option to `eo3-to-stac`, so people can tell what it's doing.
- Made schema downloads more robust
    - Use one session for all downloads
    - Use a timeout
    - Keep cached items much longer when they're stable Stac versions.
    - Allow skipping of the cache
    - Allow no-network-access usage

jeremyh committed:
Fix Travis' Python 3.6 dependencies.

Cattrs removed 3.6 support, yet it's still trying
to install the new version.

jeremyh committed:
Consistent float handling in doc comparison

jeremyh committed:
Improve, clarify stac conversion docs

whatnick reviewed (approved):
💯

whatnick reviewed (commented):
@alexgleith is better suited to review

alexgleith reviewed (approved):
Looks good to me.

As I said, it would be great to integrate this with the dataset assembler, so it's super simple for us to create STAC docs with out datasets!
